### DISCUSSION TITLE
Algebraically solve to get a closed form expression?

### DISCUSSION THREAD

**Discussion by npn0010** [casadi/casadi#3172] [Q&A] [answered]
Apologies if the answer is obvious. I've tried looking through the documentation and online forums but couldn't find anything. I am still relatively new to using CasADi also. But I am curious if there is a functionality similar to MATLAB's `solve` (https://www.mathworks.com/help/symbolic/sym.solve.html) where you can input an expression and a desired variable and it will output a closed form expression for that variable (if possible). 

For a simple example, say I have the expression `a+b=c` and I want to obtain `b=c-a`. In MATLAB this might look something like `solve(a+b==c,b)` which would return `b=c-a`.

I know this seems like a rather simple task to ask CasADi to do but I am trying to gradually move all my work entirely into a CasADi environment. Thanks in advance!

**Comment by brunoml5** [comment, marked as answer]
Hi! 

I believe that Casadi is not exactly a tool for symbolic expressions. Maybe you can work in your model in symbolic expressions and then, before calling the solvers, implement the expression in casadi variables.

Although, in the last version, casadi support "Common subexpression elimination". Take a look at the [Release Notes](https://web.casadi.org/get/).

**Reply by npn0010** [reply]
Ok good to know. Thanks! I suppose in my work flow I will still use MATLAB's symbolic toolbox and then just use CasADi for the heavy derivatives and function/MEX generation. Are there any methods for easily converting between MATLAB's symbolic variables and CasADi variables?

**Reply by brunoml5** [reply]
Hi! I don't have matlab to test it, but maybe you can generate a function file from the symbolic expression, then call the function with casadi symbols (MX or SX). Maybe that works. 

Take a look at this [tutorial ](https://www.mathworks.com/help/symbolic/generate-matlab-functions.html) from matlab. If it works, please share with us. It might be helpful for somebody else.

**Reply by npn0010** [reply]
This seems to work quite well. I was able to form a set of equations symbolically with MATLAB's toolbox, generate a `.m` function file for it with `matlabFunction`, and then input SX variables into this function and get an output in SX variables. I was also able take the Jacobian of the output using CasADi's functionality.

**Comment by jaeandersson** [comment]
There was an issue (#218) to implement such a functionality, but it was never made priority and I recently closed the issue. Typically, you want to simply have such systems of equations be solved as part of your optimization problem of simulation problem. It's also possible to create a `rootfinder` instance for solving it numerically.

**Reply by npn0010** [reply]
Maybe my use case is just kind of niche. I am solving a nonlinear root finding problem, and the reason I want to perform some algebra is to combine constraint equations and reduce the number of unknowns in the problem. This is something I just typically do by hand and then I rely on CasADi to generate the sensitivities of my root finding problem. I am really just looking for a way to combine everything into one environment to reduce chance of math errors and also just make the process of forming the problem much easier
